### DISCUSSION TITLE
Static filter syntax to only match when inline scripts are disabled

### DISCUSSION THREAD

**Discussion by jimbobmcgee** [uBlockOrigin/uBlock-issues#3427] [General]
I came to UBO from <span title="NoScript">_NSS_</span>+<span title="AdBlock Plus">_ABP_</span> a number of months ago.  

One of the significant differences I noticed after this transition was that UBO would not show the content of HTML `<noscript>` elements when it was blocking inline scripts.  NSS elects to show the contents of `<noscript>` tags where JS has been disabled, and hides these when it is enabled.

I believe I saw a closed issue at the time, stating that getting UBO to show `<noscript>` contents was either impossible or undesirable, but I can't find it now.

To compensate, I wrote myself a custom scriptlet which locates all the `<noscript>` elements in a page, writes their `innerText` to that of a new `HTMLTemplateElement` and replaces the `<noscript>` with the content of the template.  It has been more useful for me than not (e.g. for showing reminders or helper buttons when JS is disabled); in case it is at all useful for you, I have included it below.

I currently apply this scriptlet with a very broad-brush custom `*##+js(reinject-noscript)` static filter, which works as expected on every page I visit.

This, of course, now has the _opposite_ effect &ndash; when inline scripts are enabled in UBO, the contents of any `<noscript>` are _still_ shown.  There are some websites whose `<noscripts>` entirely cover the page content, or some other such annoyance, so I've taken to excluding them from my static filter, e.g. `*,~youtube.com,~youtube-nocookie.com,~accounts.google.com,~paypal.com##+js(reinject-noscript)`

I would much rather change the targeting so that UBO only applies the scriptlet when inline scripts are blocked.

Is there / could there be any syntax which could be used in a static filter, which limits the targeting to only when UBO is blocking inline scripts?  

Perhaps

* a new dollar-syntax that could apply, like `*,$no-inline-scripts##+js(reinject-noscript)`?  _(not sure of exact form; static syntax is not my strong suit!)_ 
* I don't _think_ it would be a `!#if`, because I think those occur too early in the sequence _(but happy to be proved wrong)_
* alternative approaches are also welcome...

<details>
<summary>

### Expand for custom scriptlet, as discussed...

</summary>

```js
/// reinject-noscript.js
(function reinjectNoscriptElementContent() {
  console.log('ubo.js: reinject-noscript: init');
  const fn = function() {
    console.log('ubo.js: reinject-noscript: running');
    try {
      var noscripts = document.getElementsByTagName('noscript') || [];
      for (var i = noscripts.length - 1; i >= 0; i--) {
        if (noscripts[i] !== undefined && noscripts[i] !== null) {
          var tpl = document.createElement('template');
          tpl.innerHTML = "<!-- ubo.js: reinject-noscript: start -->\n"
                        + noscripts[i].innerText
                        + "\n<!-- ubo.js: reinject-noscript: end -->";
          noscripts[i].parentNode.replaceChild(tpl.content, noscripts[i]);
        }
        else {
          console.log('ubo.js: reinject-noscript: weird = noscripts[' + i + 
                      '] not defined');
        }
      }
    }
    catch (e) {
      console.log('ubo.js: reinject-noscript: error = ' + e);
    }
  };
  if (document.readyState !== "complete") {
    document.addEventListener("DOMContentLoaded", fn);
  }
  else fn();
})();
```

</details>

**Comment by stephenhawk8054** [comment]
uBO should show the content of `noscript` if you select to wholly disable javascript: https://github.com/gorhill/uBlock/wiki/Per-site-switches#no-scripting

Can you give an example link where it doesn't work?

**Reply by garry-ut99** [reply]
> jimbobmcgee : I believe I saw a closed issue at the time, stating that getting UBO to show `<noscript>` contents was either impossible or undesirable, but I can't find it now.

Maybe one of the following:
- https://github.com/gorhill/uBlock/issues/308
- https://github.com/gorhill/uBlock/issues/1642
- https://github.com/uBlockOrigin/uBlock-issues/issues/2437
- https://github.com/uBlockOrigin/uBlock-issues/issues/2347

**Reply by jimbobmcgee** [reply]
@garry-ut99; Thanks -- that first one was indeed the one I vaguely remembered reading.  I was clearly looking in the wrong repo.

---

@stephenhawk8054; I'm sure wholly disabling works fine, but _selective_ disabling is more what I am used to.

As it stands, _for me_, the workflow for running FF with `javascript.enabled=false` or UBO with `no-scripting: * true` is a bit too painful.  I prefer to run UBO with `* * 1p-script block  * * 3p-frame block  * * 3p-script block  * * inline-script block`, because selecting from the popup matrix is far closer to the NSS approach that I came from.

My opinion (and solely an opinion, certainly not a suggestion or demand), is that the enabled state of a site's _inline scripts_ should inversely affect the visible state of its `<noscript>` tags.  I have an approach that gets me 99% of the way to that preference, and am interested to see if there is anything I have missed that would get me the rest of the way.

My thought was that there it _might_ be possible to have static filter syntax that reacted to UBO's _state_, not just the resource.  Since then, another thought was that a JS object might be made available to scriptlets (via `scriptletGlobals`...?), which contained this state.  Either of these might already exist, or might be useful enough to others to be worth implementing.  If feasible and desirable, I am open to attempting contributing it, myself.

**Reply by stephenhawk8054** [reply]
So you want to allow some scripts to run while still want `noscript` tag to be applied?

**Reply by jimbobmcgee** [reply]
Yes, usually.  UBO is a valuable front line of defence for me, but I'm happy enough to enable scripts on a case-by-case basis and also happy enough to enable some distributed functions (e.g. payment gateways) across all sites, so running with JS disabled full-time is overkill.  

But *not* seeing the content of `<noscript>` tags when I am running in my "default" state is causing me to miss certain things, and *seeing* them when running in my "relaxed" state is also causing me to make mistakes.

I'm now wondering if my preferences would be met by running `no-scripting * true; * * inline-script noop; * * 1p-script block; * * 3p-script; * * 3p-frame block` as my "default", and toggling to "relaxed" via the <kbd></></kbd> button.  i.e. nothing is allowed to run until I enable via <kbd></></kbd>, and then I'd have still control over secondary scripts.

I think it's just that my eye is drawn to the matrix on the left of the popup, and I'm missing the buttons on the right.

**Reply by stephenhawk8054** [reply]
Then it's not what the `noscript` tag is meant for. You can use the `no-scripting` toggle in the popup panel to unblock the scripts while still blocking other scripts, but the `noscript` tag will not be used any more, because you are signaling the website you are using scripts.

gorhill answered here: https://github.com/uBlockOrigin/uBlock-issues/discussions/3427#discussioncomment-11022825

**Comment by gorhill** [comment]
Declined.
